### ISSUE TITLE
Terminating: fork() called from a process already using GNU OpenMP, this is unsafe.

### ISSUE THREAD

**Issue by Accompany0313** [yezhengSTAT/scVI-3D#9]
Hi Ye，

Nice work!

I got an error:
" /MNT/second_drive/user_data ygc/local/lib/python3.8 / site - packages/sklearn/utils/deprecation py: 86: FutureWarning: Function setup_anndata is deprecated;  Please use the model-specific setup_anndata methods instead. The global method will be removed in version 0.15.0.
warnings.warn(msg, category=FutureWarning)
INFO     Using batches from adata.obs["batch"]
INFO     No label_key inputted, assuming all cells have same label
INFO     Using data from adata.X
Terminating: fork() called from a process already using GNU OpenMP, this is unsafe.
/ MNT/second_drive/user_data ygc/local/lib/python3.8 / site - packages/sklearn/utils/deprecation py: 86: FutureWarning: Function setup_anndata is deprecated;  Please use the model-specific setup_anndata methods instead. The global method will be removed in version 0.15.0.
warnings.warn(msg, category=FutureWarning)
INFO     Using batches from adata.obs["batch"]
INFO     No label_key inputted, assuming all cells have same label
INFO     Using data from adata.X
Terminating: fork() called from a process already using GNU OpenMP, this is unsafe."

I have no idea what could cause the error. The command I use is:
python scripts/scVI-3D.py -b 10 -c "whole" -r 1000000 -i "demoData" -o "results" -cs "supplementaryData/demoData_summary.txt" 
-g "supplementaryData/hg19.chrom.sizes" -br -n 100 -gpu -p 10 -pca 50 -up -tp -v

Could you please help me with it? Thank you so much!

**Comment by yezhengSTAT** [comment]
Hello,
It seems that the newer version of sklearn package that you probably is using has deprecated setup_anndata function. I think we developed the package using 0.24.2 version of scikit-learn (for other python module, the version we used can be found at https://github.com/yezhengSTAT/scVI-3D#1-preparation). Sorry for the trouble as we did not expect the new dependent tool is not compatible with its old version. 

Thanks,
Ye

**Comment by Accompany0313** [comment]
Hi Ye,

Thanks a lot for your timely response! 

I changed the version of scikit-learn to 0.24.2, but I still have the following issues：
/home/ygc/.conda/envs/scvi-3d/lib/python3.9/site-packages/sklearn/utils/deprecation.py:86: FutureWarning: Function setup_anndata is deprecated; Please use the model-specific setup_anndata methods instead. The global method will be removed in version 0.15.0.
  warnings.warn(msg, category=FutureWarning)
INFO     Using batches from adata.obs["batch"]                                                                                                                                       
INFO     No label_key inputted, assuming all cells have same label                                                                                                                   
INFO     Using data from adata.X                                                                                                                                                     
/home/ygc/.conda/envs/scvi-3d/lib/python3.9/abc.py:119: FutureWarning: SparseDataset is deprecated and will be removed in late 2024. It has been replaced by the public classes CSRDataset and CSCDataset.

For instance checks, use `isinstance(X, (anndata.experimental.CSRDataset, anndata.experimental.CSCDataset))` instead.

For creation, use `anndata.experimental.sparse_dataset(X)` instead.

  return _abc_instancecheck(cls, instance)
Terminating: fork() called from a process already using GNU OpenMP, this is unsafe.

The conda environment I use is ：
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                       2_gnu    conda-forge
absl-py                   2.1.0                    pypi_0    pypi
aiohttp                   3.9.3                    pypi_0    pypi
aiosignal                 1.3.1                    pypi_0    pypi
anndata                   0.10.6                   pypi_0    pypi
array-api-compat          1.5.1                    pypi_0    pypi
asttokens                 2.4.1                    pypi_0    pypi
async-timeout             4.0.3                    pypi_0    pypi
attrs                     23.2.0                   pypi_0    pypi
ca-certificates           2024.3.11            h06a4308_0  
comm                      0.2.2                    pypi_0    pypi
contourpy                 1.2.0                    pypi_0    pypi
cycler                    0.12.1                   pypi_0    pypi
decorator                 5.1.1                    pypi_0    pypi
dnspython                 2.6.1                    pypi_0    pypi
docrep                    0.3.2                    pypi_0    pypi
et-xmlfile                1.1.0                    pypi_0    pypi
exceptiongroup            1.2.0                    pypi_0    pypi
executing                 2.0.1                    pypi_0    pypi
filelock                  3.13.3                   pypi_0    pypi
fonttools                 4.50.0                   pypi_0    pypi
frozenlist                1.4.1                    pypi_0    pypi
fsspec                    2024.3.1                 pypi_0    pypi
future                    1.0.0                    pypi_0    pypi
get-annotations           0.1.2                    pypi_0    pypi
grpcio                    1.62.1                   pypi_0    pypi
h5py                      3.10.0                   pypi_0    pypi
hyperopt                  0.1.2                    pypi_0    pypi
idna                      3.6                      pypi_0    pypi
importlib-metadata        7.1.0                    pypi_0    pypi
importlib-resources       6.4.0                    pypi_0    pypi
ipython                   8.18.1                   pypi_0    pypi
ipywidgets                8.1.2                    pypi_0    pypi
jedi                      0.19.1                   pypi_0    pypi
jinja2                    3.1.3                    pypi_0    pypi
joblib                    1.3.2                    pypi_0    pypi
jupyterlab-widgets        3.0.10                   pypi_0    pypi
kiwisolver                1.4.5                    pypi_0    pypi
ld_impl_linux-64          2.38                 h1181459_1  
libffi                    3.4.4                h6a678d5_0  
libgcc-ng                 13.2.0               h807b86a_5    conda-forge
libgomp                   13.2.0               h807b86a_5    conda-forge
libstdcxx-ng              11.2.0               h1234567_1  
lightning-utilities       0.11.1                   pypi_0    pypi
llvmlite                  0.42.0                   pypi_0    pypi
markdown                  3.6                      pypi_0    pypi
markdown-it-py            3.0.0                    pypi_0    pypi
markupsafe                2.1.5                    pypi_0    pypi
matplotlib                3.8.3                    pypi_0    pypi
matplotlib-inline         0.1.6                    pypi_0    pypi
mdurl                     0.1.2                    pypi_0    pypi
mpmath                    1.3.0                    pypi_0    pypi
multidict                 6.0.5                    pypi_0    pypi
natsort                   8.4.0                    pypi_0    pypi
ncurses                   6.4                  h6a678d5_0  
networkx                  3.2.1                    pypi_0    pypi
numba                     0.59.1                   pypi_0    pypi
numpy                     1.26.4                   pypi_0    pypi
nvidia-cublas-cu12        12.1.3.1                 pypi_0    pypi
nvidia-cuda-cupti-cu12    12.1.105                 pypi_0    pypi
nvidia-cuda-nvrtc-cu12    12.1.105                 pypi_0    pypi
nvidia-cuda-runtime-cu12  12.1.105                 pypi_0    pypi
nvidia-cudnn-cu12         8.9.2.26                 pypi_0    pypi
nvidia-cufft-cu12         11.0.2.54                pypi_0    pypi
nvidia-curand-cu12        10.3.2.106               pypi_0    pypi
nvidia-cusolver-cu12      11.4.5.107               pypi_0    pypi
nvidia-cusparse-cu12      12.1.0.106               pypi_0    pypi
nvidia-nccl-cu12          2.19.3                   pypi_0    pypi
nvidia-nvjitlink-cu12     12.4.99                  pypi_0    pypi
nvidia-nvtx-cu12          12.1.105                 pypi_0    pypi
openpyxl                  3.1.2                    pypi_0    pypi
openssl                   3.0.13               h7f8727e_0  
opt-einsum                3.3.0                    pypi_0    pypi
packaging                 24.0                     pypi_0    pypi
pandas                    2.2.1                    pypi_0    pypi
parso                     0.8.3                    pypi_0    pypi
patsy                     0.5.6                    pypi_0    pypi
pexpect                   4.9.0                    pypi_0    pypi
pillow                    10.2.0                   pypi_0    pypi
pip                       23.3.1           py39h06a4308_0  
prompt-toolkit            3.0.43                   pypi_0    pypi
protobuf                  5.26.0                   pypi_0    pypi
ptyprocess                0.7.0                    pypi_0    pypi
pure-eval                 0.2.2                    pypi_0    pypi
pydeprecate               0.3.0                    pypi_0    pypi
pygments                  2.17.2                   pypi_0    pypi
pymongo                   4.6.2                    pypi_0    pypi
pynndescent               0.5.11                   pypi_0    pypi
pyparsing                 3.1.2                    pypi_0    pypi
pyro-api                  0.1.2                    pypi_0    pypi
pyro-ppl                  1.9.0                    pypi_0    pypi
python                    3.9.19               h955ad1f_0  
python-dateutil           2.9.0.post0              pypi_0    pypi
pytorch-lightning         1.3.8                    pypi_0    pypi
pytz                      2024.1                   pypi_0    pypi
pyyaml                    5.4.1                    pypi_0    pypi
readline                  8.2                  h5eee18b_0  
resampy                   0.3.1                    pypi_0    pypi
rich                      13.7.1                   pypi_0    pypi
scanpy                    1.9.8                    pypi_0    pypi
scikit-learn              0.24.2                   pypi_0    pypi
scikit-misc               0.3.1                    pypi_0    pypi
scipy                     1.12.0                   pypi_0    pypi
scvi-tools                0.14.5                   pypi_0    pypi
seaborn                   0.13.2                   pypi_0    pypi
session-info              1.0.0                    pypi_0    pypi
setuptools                68.2.2           py39h06a4308_0  
six                       1.16.0                   pypi_0    pypi
sqlite                    3.41.2               h5eee18b_0  
stack-data                0.6.3                    pypi_0    pypi
statsmodels               0.14.1                   pypi_0    pypi
stdlib-list               0.10.0                   pypi_0    pypi
style                     1.1.0                    pypi_0    pypi
sympy                     1.12                     pypi_0    pypi
tensorboard               2.16.2                   pypi_0    pypi
tensorboard-data-server   0.7.2                    pypi_0    pypi
threadpoolctl             3.4.0                    pypi_0    pypi
tk                        8.6.12               h1ccaba5_0  
torch                     2.2.1                    pypi_0    pypi
torchmetrics              0.5.0                    pypi_0    pypi
tqdm                      4.66.2                   pypi_0    pypi
traitlets                 5.14.2                   pypi_0    pypi
triton                    2.2.0                    pypi_0    pypi
typing-extensions         4.10.0                   pypi_0    pypi
tzdata                    2024.1                   pypi_0    pypi
umap-learn                0.5.5                    pypi_0    pypi
update                    0.0.1                    pypi_0    pypi
wcwidth                   0.2.13                   pypi_0    pypi
werkzeug                  3.0.1                    pypi_0    pypi
wheel                     0.41.2           py39h06a4308_0  
widgetsnbextension        4.0.10                   pypi_0    pypi
xlrd                      2.0.1                    pypi_0    pypi
xz                        5.4.6                h5eee18b_0  
yarl                      1.9.4                    pypi_0    pypi
zipp                      3.18.1                   pypi_0    pypi
zlib                      1.2.13               h5eee18b_0  

Thank you so much!

**Comment by yezhengSTAT** [comment]
The version compatibility issue indeed needs some further adjustment work, but I guess that will only give us warning messages. Terminating errors related to fork() and GNU OpenMP is the critical error stopping your process. The error message "Terminating: fork() called from a process already using GNU OpenMP, this is unsafe" suggests that your script or one of the libraries it uses tries to use multiprocessing in a way that's not safe or compatible with OpenMP (Open Multi-Processing), which is used for parallel programming in C/C++ and Fortran. This error can occur when a process that's already running in a parallelized manner (using OpenMP) attempts to spawn new processes (via fork() in Python). This is often unsafe because it can lead to deadlocks or other concurrency issues. Maybe try to use one thread by setting OMP_NUM_THREADS=1?

**Comment by Accompany0313** [comment]
Thank you very much for your advice. I tried to set OMP_NUM_THREADS=1 first, but the same problem still occurred. I tried again to set the number of cpus to 1, and this time the problem was resolved. However, the running time has been extended from 15 minutes to more than two hours. Finally, I would like to express my sincere thanks for your reply.

The command I use is:
python scripts/scVI-3D.py -b 10 -c "whole" -r 1000000 -i "demoData" -o "results" -cs "supplementaryData/demoData_summary.txt"
-g "supplementaryData/hg19.chrom.sizes" -br -n 100 -gpu -p 1 -pca 50 -up -tp -v
